### PR TITLE
Antismashlite download databases

### PR DESCRIPTION
**Pull request by jasmezz**
This module downloads the antiSMASH databases (using the [lite version](https://quay.io/repository/biocontainers/antismash-lite) of antiSMASH).

## PR checklist

Closes #XXX 

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

File: modules/antismash/antismashlitedownloaddatabases/main.nf [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    tag '$database_dir'
+    label 'process_low'
+
+    conda (params.enable_conda ? "bioconda::antismash-lite=6.0.1" : null)
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/antismash-lite:6.0.1--pyhdfd78af_0' :
+        'quay.io/biocontainers/antismash-lite:6.0.1--pyhdfd78af_0' }"
+
+    input:
+    val(database_dir)
```
Comment (jfy133) [inline modules/antismash/antismashlitedownloaddatabases/main.nf:11]: This confuses me, the database isn't download yet, are you specifying 'where' the database should be downloaded?

I wouldn't do that - if necessary hardcode it into a dir name (e.g. antismash-dbs or something). NXF should deal with the location of internally downloaded files, and the `modules.config` file publishDir options can allow a user tosave this somewhere for re-use.

Alternative if this is necessary, use `prefix` instead

File: modules/antismash/antismashlitedownloaddatabases/main.nf [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+    output:
+    path(database_dir) , emit: database
+
```
Comment (jfy133) [inline modules/antismash/antismashlitedownloaddatabases/main.nf:15]:
```suggestion
```

File: modules/antismash/antismashlitedownloaddatabases/main.nf [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    conda (params.enable_conda ? "bioconda::antismash-lite=6.0.1" : null)
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/antismash-lite:6.0.1--pyhdfd78af_0' :
+        'quay.io/biocontainers/antismash-lite:6.0.1--pyhdfd78af_0' }"
+
+    input:
+    val(database_dir)
+
+    output:
+    path(database_dir) , emit: database
```
Comment (jfy133) [inline modules/antismash/antismashlitedownloaddatabases/main.nf:14]: This probably woudn't work, as you would have a conflict between the input/output varaible (NXF wouldn't know which to pick up... they can't be the same file.

File: modules/antismash/antismashlitedownloaddatabases/meta.yml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+      homepage: https://docs.antismash.secondarymetabolites.org
+      documentation: https://docs.antismash.secondarymetabolites.org
+      tool_dev_url: https://github.com/antismash/antismash
+      doi: "10.1093/nar/gkab335"
+      licence: ['AGPL v3']
+
+input:
+  - database_dir:
+      type: directory
+      description: Directory containing the user-specified path for the download of antiSMASH databases
```
Comment (jfy133) [inline modules/antismash/antismashlitedownloaddatabases/meta.yml:29]: See question above

File: modules/antismash/antismashlitedownloaddatabases/meta.yml [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+output:
+  - versions:
+      type: file
+      description: File containing software versions
+      pattern: "versions.yml"
+
+  - database:
+      type: directory
+      description: Downloaded clusterblast databases
+      pattern: "lib/python3.8/site-packages/antismash/databases/"
```
Comment (jfy133) [inline modules/antismash/antismashlitedownloaddatabases/meta.yml:41]: is this correct? Above you were saying this location was a user defined location?

File: tests/modules/antismash/antismashlitedownloaddatabases/test.yml [hunk 6]
```diff
@@ -0,0 +1,8 @@
+- name: antismash antismashlitedownloaddatabases test_antismash_antismashlitedownloaddatabases
+  command: nextflow run tests/modules/antismash/antismashlitedownloaddatabases -entry test_antismash_antismashlitedownloaddatabases -c tests/config/nextflow.config
+  tags:
+    - antismash/antismashlitedownloaddatabases
+    - antismash
+  files:
+    - path: output/antismash/versions.yml
+      md5sum: e2656c8d2bcc7469eba40eb1ee5c91b3
```
Comment (jfy133) [inline tests/modules/antismash/antismashlitedownloaddatabases/test.yml:8]: Missing the actual DB path?

can be simply that it requirs `-path: /<oath>/<to>/<dir>`, don't need a md5sum (as not psossible for directories anyway)

**Review by jfy133 — CHANGES_REQUESTED** [review]
Overall looks OK, I'm not sure about what the input parameter is going on here and you're missing a working test (as it doesn't include the actual db atm?)

File: modules/antismash/antismashlitedownloaddatabases/main.nf (hunk 1 above)
Comment (jasmezz) [inline modules/antismash/antismashlitedownloaddatabases/main.nf:11]:
```suggestion
```

File: modules/antismash/antismashlitedownloaddatabases/main.nf (hunk 3 above)
Comment (jasmezz) [inline modules/antismash/antismashlitedownloaddatabases/main.nf:14]:
```suggestion
    path("antismash_db") , emit: database
```

File: modules/antismash/antismashlitedownloaddatabases/main.nf [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def databases = database_dir ? "--database-dir ${database_dir}" : ''
+    """
+    download-antismash-databases \\
+        $databases \\
```
Comment (jasmezz) [inline modules/antismash/antismashlitedownloaddatabases/main.nf:26]:
```suggestion
        --database-dir antismash_db \\
```

File: modules/antismash/antismashlitedownloaddatabases/meta.yml [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+      tool_dev_url: https://github.com/antismash/antismash
+      doi: "10.1093/nar/gkab335"
+      licence: ['AGPL v3']
+
+input:
+  - database_dir:
+      type: directory
+      description: Directory containing the user-specified path for the download of antiSMASH databases
+      pattern: "*"
+
```
Comment (jasmezz) [inline modules/antismash/antismashlitedownloaddatabases/meta.yml:31]:
```suggestion
```

File: modules/antismash/antismashlitedownloaddatabases/main.nf [hunk 9]
```diff
@@ -0,0 +1,34 @@
+process ANTISMASH_ANTISMASHLITEDOWNLOADDATABASES {
+    tag '$database_dir'
+    label 'process_low'
+
+    conda (params.enable_conda ? "bioconda::antismash-lite=6.0.1" : null)
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/antismash-lite:6.0.1--pyhdfd78af_0' :
+        'quay.io/biocontainers/antismash-lite:6.0.1--pyhdfd78af_0' }"
+
```
Comment (jasmezz) [inline modules/antismash/antismashlitedownloaddatabases/main.nf:9]:
```suggestion
```

File: modules/antismash/antismashlitedownloaddatabases/meta.yml (hunk 5 above)
Comment (jasmezz) [inline modules/antismash/antismashlitedownloaddatabases/meta.yml:41]:
```suggestion
      description: Download directory for antiSMASH databases
      pattern: "antismash_db"
```

File: tests/modules/antismash/antismashlitedownloaddatabases/main.nf [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+
+nextflow.enable.dsl = 2
+
+include { ANTISMASH_ANTISMASHLITEDOWNLOADDATABASES } from '../../../../modules/antismash/antismashlitedownloaddatabases/main.nf'
+
+workflow test_antismash_antismashlitedownloaddatabases {
+
+    input = ''
+
+    ANTISMASH_ANTISMASHLITEDOWNLOADDATABASES ( input )
```
Comment (jasmezz) [inline tests/modules/antismash/antismashlitedownloaddatabases/main.nf:11]:
```suggestion
    ANTISMASH_ANTISMASHLITEDOWNLOADDATABASES ( )
```

File: tests/modules/antismash/antismashlitedownloaddatabases/nextflow.config [hunk 11]
```diff
@@ -0,0 +1,5 @@
+process {
+
+    publishDir = { "${params.outdir}/${task.process.tokenize(':')[-1].tokenize('_')[0].toLowerCase()}" }
+    
+}
```
Comment (jasmezz) [inline tests/modules/antismash/antismashlitedownloaddatabases/nextflow.config:5]:
```suggestion
}

```

File: tests/modules/antismash/antismashlitedownloaddatabases/test.yml [hunk 12]
```diff
@@ -0,0 +1,10 @@
...
+    - path: output/antismash/antismash_db
```
Comment (Midnighter) [inline tests/modules/antismash/antismashlitedownloaddatabases/test.yml:9]: I would maybe test for the presence of a few more expected directories/files.

**Review by Midnighter — APPROVED** [review]
Looks good to me! I haven't checked why the tests are failing but we can talk about that tomorrow.

I was wondering a little, though, if it's strictly necessary to have the download in nf-core modules rather than just as a local module in the funcscan pipeline... Did you discuss that at some point?

**Comment by Midnighter** [comment]
So mounting the correct directory structure did not work out? Having such a long command might actually break things.

**Review by jfy133 — COMMENTED** [review]
In terms of modules guidelines wise, this looks OK now, however it comes across as very circular - your input is channels what you're downloading?

```
      description: antismash/outputs/html/css folder which is being created during the antiSMASH database downloading step
```

Is that correct?

I also agree with @Midnighter we should add more test paths, to check for but I think this is would otherwise be OK.

**Review by jfy133 — APPROVED** [review]
Once `meta.yml` is updated for describing the input channels are circular and this module is NOT recommended for pipeline use

Add a few more output files checks in the test.yml

File: modules/antismash/antismashlitedownloaddatabases/meta.yml [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+      homepage: https://docs.antismash.secondarymetabolites.org
+      documentation: https://docs.antismash.secondarymetabolites.org
+      tool_dev_url: https://github.com/antismash/antismash
+      doi: "10.1093/nar/gkab335"
+      licence: ["AGPL v3"]
+
+input:
+  - database_css:
+      type: directory
+      description: antismash/outputs/html/css folder which is being created during the antiSMASH database downloading step
```
Comment (jfy133) [inline modules/antismash/antismashlitedownloaddatabases/meta.yml:29]:
```suggestion
      description: |
      antismash/outputs/html/css folder which is being created during the antiSMASH database downloading step. This is required due to particularities of antiSMASH that is not very compatible with container engines. These files are normally downloaded by download-antismash-databases itself, and must be retrieved by the use by manually running the command with conda or a standalone installation of antiSMASH. Therefore we do not recommend using this module for production pipelines, but rather require users to specify their own local copy of the antiSMASH database in pipelines.
```

File: modules/antismash/antismashlitedownloaddatabases/meta.yml [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+      pattern: "css"
+  - database_detection:
+      type: directory
+      description: antismash/detection folder which is being created during the antiSMASH database downloading step
+      pattern: "detection"
+  - database_modules:
+      type: directory
+      description: antismash/modules folder which is being created during the antiSMASH database downloading step
```
Comment (jfy133) [inline modules/antismash/antismashlitedownloaddatabases/meta.yml:37]:
```suggestion
      description: |
      These files are normally downloaded by download-antismash-databases itself, and must be retrieved by the use by manually running the command with conda or a standalone installation of antiSMASH. Therefore we do not recommend using this module for production pipelines, but rather require users to specify their own local copy of the antiSMASH database in pipelines.
```

File: modules/antismash/antismashlitedownloaddatabases/main.nf [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+        'https://depot.galaxyproject.org/singularity/antismash-lite:6.0.1--pyhdfd78af_1' :
+        'quay.io/biocontainers/antismash-lite:6.0.1--pyhdfd78af_1' }"
+    containerOptions {
+        workflow.containerEngine == 'singularity' ?
+        "-B $database_css:/usr/local/lib/python3.8/site-packages/antismash/outputs/html/css,$database_detection:/usr/local/lib/python3.8/site-packages/antismash/detection,$database_modules:/usr/local/lib/python3.8/site-packages/antismash/modules" :
+        workflow.containerEngine == 'docker' ?
+        "-v \$PWD/$database_css:/usr/local/lib/python3.8/site-packages/antismash/outputs/html/css -v \$PWD/$database_detection:/usr/local/lib/python3.8/site-packages/antismash/detection -v \$PWD/$database_modules:/usr/local/lib/python3.8/site-packages/antismash/modules" :
+        ''
+        }
+    input:
```
Comment (jfy133) [inline modules/antismash/antismashlitedownloaddatabases/main.nf:15]:
```suggestion
    
    input:
```

File: modules/antismash/antismashlitedownloaddatabases/main.nf [hunk 16]
```diff
@@ -0,0 +1,39 @@
+process ANTISMASH_ANTISMASHLITEDOWNLOADDATABASES {
+    label 'process_low'
+
+    conda (params.enable_conda ? "bioconda::antismash-lite=6.0.1" : null)
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/antismash-lite:6.0.1--pyhdfd78af_1' :
+        'quay.io/biocontainers/antismash-lite:6.0.1--pyhdfd78af_1' }"
+    containerOptions {
```
Comment (jfy133) [inline modules/antismash/antismashlitedownloaddatabases/main.nf:8]:
```suggestion

    /* 
    These files are normally downloaded by download-antismash-databases itself, and must be retrieved for input by manually running the command with conda or a standalone installation of antiSMASH. Therefore we do not recommend using this module for production pipelines, but rather require users to specify their own local copy of the antiSMASH database in pipelines. This is solely for use for CI tests of the nf-core/module version of antiSMASH
    Reason: the tool on execution tries to modify contents of database files stored within the container itself based on the order of files according to their timestamp. During container construction these timestamps get incorrectly modified and the re-genration of the files within the container is attempted. By mounting the files it prevents the tool from trying to modify database files. 
    */
    
    containerOptions {
```

Or something...

**Review by jfy133 — APPROVED** [review]
DOCS! DOCS EVERYWHERE!

**Comment by jfy133** [comment]
I approve merging now - singularity was working previously, currently biocontainer is down so the image can't be pulled.

**Pull request merged**